man/minbif.8
.TH MINBIF 8 "04 December 2011"
.SH NAME
minbif \- The IRC instant messaging gateway
.SH SYNOPSIS
.hy 0
\fBminbif [\-hv] [\-\-pidfile \fIpidfile\fB]\fP
.I CONFIG_FILE
.SH DESCRIPTION
.LP
Minbif is an IRC gateway to IM networks which provides the following features:

.nf
* Minbif uses a library which abstracts all IM calls, and has several plugins to support more than 15 IM protocols (IRC included!);
* Two modes: inetd and daemon fork;
* Only IRC commands are used to control Minbif;
* IM Certificates check;
* Buddies are IRC users;
* Each account is associated to a status channel, where all connected buddies are;
* Add and remove buddies from list with /INVITE and /KICK commands;
* Blocked users are bans on the account's status channel;
* Display when a buddy is typing a message;
* Can chat with someone who is not in your buddy list;
* You can see buddies' icons (with libcaca) or download them;
* DCC SEND an image to set your icon on IM networks;
* Display extended information about buddies with /WII command;
* Status management;
* Can send and receive files, which are sent or received to/from your IRC client with DCC SEND;
* Conversation channels are supported;
* Auto\-rejoin conversation channels at connection;
* Display list of channels on an IM account with /LIST;
* irssi scripts to increase your user experience of minbif;
* CACAcam (webcam in ascii art);
* CoinCoin plugin for libpurple;
* PAM support;
* TLS support with certificate auth.
.fi

.SH OPTIONS
.TP
\fB\-\-pidfile\fR \fIpidfile\fR
path to pidfile where the minbif's PID is written.
.TP
\fB\-\-version\fR
display version of minbif.
.TP
\fB\-\-help\fR
show help.
.TP
.B config_file
Configuration file location.

.SH CONFIGURATION
Edit the \fIminbif.conf\fP file and change value when needed.

With the inetd mode, use this command to add minbif:
.nf
# update\-inetd \-\-add '56667 stream tcp nowait minbif /usr/sbin/tcpd /usr/bin/minbif /etc/minbif/minbif.conf'
.fi

With the daemon fork mode, run minbif with:
.nf
# minbif /etc/minbif/minbif.conf
.fi

.SH HOW TO USE
Connect your IRC client (for examble \fIirssi\fP) on minbif with this command:
.nf
/server localhost 56667 \fIpassword\fP
.fi

The nickname set on your IRC client is your username on minbif. First time you
connect to minbif with this nickname, the account is created and the
\fIpassword\fP parameter is set.

Next, use the \fB/map help\fP command to know how to add a network.

See also: http://minbif.im/Quick_start

.SH COPYRIGHT
Copyright(C) 2009-2011 Romain Bignon
.LP
For full COPYRIGHT see COPYING file with minbif package.
.LP
.RE
.SH FILES
 "minbif.conf"


ChangeLog
Minbif 1.0.5 (2011-12-04)
	* Add an option 'file_transfers/dcc_own_ip' in config file.
	* Change weechat's script to ignore the 'request' nick.
	* When a new request is received during process of an other one, tell
	  user.
	* Add a facebook rename script for weechat.
	* Change users of child process with pam authentification.
	* Escape URL entities in /JOIN params.
	* Add a /SCONNECT alias to /CONNECT.
	* Can use wildcards to match buddies with /WHO.
	* prpl-coincoin: support hh:mm:ss:ref norloge syntax.
	* prpl-coincoin: works with the new version of the DLFP board.
	* Fix: prevent crash with the msn plugin.
	* Fix: conversion to IRC escape chars.
	* Fix: compatibility with libpurple 2.8+.

Minbif 1.0.4 (2010-09-20)
	* libpurple 2.4 isn't supported anymore.
	* /WHO can take -s to display status instead of realname.
	* Support multilines messages by waiting a delay before sending buffer
	  (#195).
	* Send RPL_MYINFO and RPL_ISUPPORT at connection to be RFC 2812
	  compliant.
	* Only /SVSNICK sends aliases server side.
	* New command '/STATS p' to manage libpurple plugins.
	* /MAP commands have to match the start of the exact name.
	* With '/MAP add', decode usernames as URL-encoded (#338).
	* Do not add '0' at the end of the new account ID when it's the first
	  one.
	* Allow joining a remote channel even if account is disconnected
	  (#387).
	* New option '-m' to binary to select the server mode to create.
	* Add im_typing_notice script for weechat.
	* prpl-coincoin new features: support clocks without seconds, can
	  refer to a previous message by writing the nickname, etc.
	* Fix: crash when which occurs sometimes when disconnecting from an
	  account.
	* Fix: don't warn about aborted joins if the account is not connected
	  while removing it (#337).
	* Fix: issues with multilines received messages (#331).
	* Fix: use unicode in status channel topics.
	* Fix: ability to unset status.
	* Fix: file descriptor leak in daemon fork mode.
	* Fix: infinite loop when there are more than 2 accounts without any
	  ID set.

Minbif 1.0.3 (2010-05-15)
	* Use /topic on status channels to change your status message (#210).
	* With field requests, display image fields with libcaca.
	* Improvement of bans management (match wildcards, fix parsing, etc.).
	* When starting to chat with a buddy in a status channel, every
	  received messages from this buddy are displayed in the status channel
	  (#179).
	* Do not allow two IRC clients to be logged at the same time on the
	  same account (only with the daemon fork mode) (#178).
	* Display the current status of a buddy in /WHOIS when he is not away.
	* Default value of 'server_alias' is now 'true'.
	* Works with libpurple 2.7.0.
	* Fix: crash with conversations (occured when a buddy sends a message,
	  then disconnects/reconnects and re-send a message) (#258).
	* Fix: crash when closing a request.
	* Fix: compilation on FreeBSD (now using a buildbot slave on this OS)
	  (#255).
	* Fix: /svsnick collision detection.
	* Fix: crash when nobody connects to DCC server.
	* Fix: crash when a buddy has an icon without any path (wtf).

Minbif 1.0.2 (2010-04-05)
	* TLS support.
	* New 'maxcon' parameter to limit simultaneous connections.
	* /TOPIC on a status channel change your status on attached accounts.
	* Can put '%XX' in remote channel names to join, to insert special
	  characters (for example spaces).
	* Auto-reconnect on accounts only when disconnection was because of a
	  network error.
	* Display account ID and more information in requests.
	* Aibility to not store password on an account (it is requested at
	  connection).
	* Aibility to edit password with '/MAP edit'.
	* Use imlib2 to convert received buddyicons to right encoded images.
	* Send several PRIVMSG or NOTICE when there are \n in message.
	* Better display of /MAP replies (with stats on accounts).
	* /ADMIN voiced_buddies: enable or disable voices on status channels.
	* /ADMIN accept_nobuddies_messages: if disabled, messages from unknown
	  buddies are not displayed (prevent spam).
	* /ADMIN server_aliases: enable or disable sending of aliases
	  server-side.
	* 'server_aliases' accounts parameter to enable or disable sending of
	  aliases server-side on a specific account.
	* Command '/STATS u' to display server uptime.
	* Command '/STATS o' to display every available minbif administrator
	  and their email address.
	* Command '/INFO' to display copyright information.
	* Irssi script to auto-rename facebook buddies.
	* Fix: pam default configuration.
	* Fix: chat buddy's nickname when users have a jabber resource.
	* Fix: crash with '/STATS c' on a disconnected account.
	* Fix: compilation on strange compilers.
	* Fix: typing notices were broken.
	* Fix: crash when removing an account which have an active request.
	* Fix: prevent empty nicknames.
	* Fix: sent buddy icons cut on the bottom.

Minbif 1.0.1 (2010-01-16)
	* PAM authentication.
	* /MAP CMD command to run a specific command on an account, which
	  allow you to change nickname, mood messages, etc.
	* Support fields requests.
	* Display group name in IRC user realname.
	* The /MAP REGISTER command registers account on server before adding
	  it.
	* Conf option to enable the purple conv logging.
	* /CMD command to send commands in a conversation
	* Display date before delayed messages if they hasn't been sent today.
	* Fix: daemonize correctly.
	* Fix: security issue when a global password is set on the minbif
	  server, it creates the userdir before checking if the password is
	  correct.
	* Fix: buddy's realnames wasn't updated in realtime.
	* Fix: the PONG reply wasn't RFC compliant.
	* Fix: crash in NAMES and TOPIC commands when giving an invalid chan
	  name.
	* Fix: don't typing notice unless either bitlbee_send_typing or
	  typing_notice is true.

Minbif 1.0 (2009-11-18)
	* Global proxy settings with the /ADMIN command.
	* Ability to set parameters when joining a conversation channel.

src/core/minbif.cpp
/*
 * Minbif - IRC instant messaging gateway
 * Copyright(C) 2009-2011 Romain Bignon, Marc Dequènes (Duck)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <cerrno>
#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <cstring>
#include <sys/resource.h>
#include <getopt.h>
#include <sys/types.h>
#include <unistd.h>

#include "minbif.h"
#include "sighandler.h"
#include "version.h"
#include "log.h"
#include "util.h"
#include "im/im.h"
#include "server_poll/poll.h"

const char* infotxt[] = {
	MINBIF_VERSION,
	strdup(MINBIF_BUILD.c_str()), /* don't care if we don't free it as it can be used until exit. */
	"",
	"Copyright(C) 2009-2011 Romain Bignon",
	"This program is free software; you can redistribute it and/or modify",
	"it under the terms of the GNU General Public License as published by",
	"the Free Software Foundation, version 2 of the License.",
	NULL
};

Minbif::Minbif()
	: loop(NULL),
	  server_poll(0)
{
	ConfigSection* section;
	ConfigSection* sub;

	section = conf.AddSection("path", "Path information", MyConfig::NORMAL);
	section->AddItem(new ConfigItem_string("users", "Users directory"));
	section->AddItem(new ConfigItem_string("motd", "Path to motd", " "));

	section = conf.AddSection("irc", "Server information", MyConfig::NORMAL);
	section->AddItem(new ConfigItem_string("hostname", "Server hostname", " "));
	section->AddItem(new ConfigItem_string("password", "Global server password", " "));
	section->AddItem(new ConfigItem_int("type", "Type of daemon", 0, 2, "0"));
	section->AddItem(new ConfigItem_int("ping", "Ping frequence (s)", 0, 65535, "60"));
	section->AddItem(new ConfigItem_string("buddy_icons_url", "URL to display in /WHOIS to get a buddy icon", " "));

	sub = section->AddSection("inetd", "Inetd information", MyConfig::OPTIONAL);
	add_server_block_common_params(sub);

	sub = section->AddSection("daemon", "Daemon information", MyConfig::OPTIONAL);
	sub->AddItem(new ConfigItem_string("bind", "IP address to listen on"));
	sub->AddItem(new ConfigItem_int("port", "Port to listen on", 1, 65535), true);
	sub->AddItem(new ConfigItem_bool("background", "Start minbif in background", "true"));
	sub->AddItem(new ConfigItem_int("maxcon", "Maximum simultaneous connections", 0, 65535, "0"));
	add_server_block_common_params(sub);

	sub = section->AddSection("oper", "Define an IRC operator", MyConfig::MULTIPLE);
	sub->AddItem(new ConfigItem_string("login", "Nickname of IRC operator"), true);
	sub->AddItem(new ConfigItem_string("password", "IRC operator password"));
	sub->AddItem(new ConfigItem_string("email", "IRC operator email address", "*@*"));

	section = conf.AddSection("aaa", "Authentication, Authorization and Accounting", MyConfig::OPTIONAL);
	section->AddItem(new ConfigItem_bool("use_local", "Use local database to authenticate users", "true"));
#ifdef HAVE_PAM
	section->AddItem(new ConfigItem_bool("use_pam", "Use PAM mechanisms to authenticate/authorize users", "false"));
	section->AddItem(new ConfigItem_bool("pam_setuid", "Child process setuid with the pam user (needs root and pam auth)", "false"));
#endif
	section->AddItem(new ConfigItem_bool("use_connection", "Use connection information to authenticate/authorize users", "false"));

	section = conf.AddSection("file_transfers", "File transfers parameters", MyConfig::OPTIONAL);
	section->AddItem(new ConfigItem_bool("enabled", "Enable file transfers", "true"));
	section->AddItem(new ConfigItem_bool("dcc", "Send files to IRC user with DCC", "true"));
	section->AddItem(new ConfigItem_string("dcc_own_ip", "Force minbif to always send DCC requests from a particular IP address", " "));
	section->AddItem(new ConfigItem_intrange("port_range", "Port range to listen on for DCC", 1024, 65535, "1024-65535"));

	section = conf.AddSection("logging", "Log information", MyConfig::NORMAL);
	section->AddItem(new ConfigItem_string("level", "Logging level"));
	section->AddItem(new ConfigItem_bool("to_syslog", "Log error and warnings to syslog"));
	section->AddItem(new ConfigItem_bool("conv_logs", "Enable conversation logging", "false"));

}

void Minbif::add_server_block_common_params(ConfigSection* section)
{
	section->AddItem(new ConfigItem_string("security", "none/tls/starttls/starttls-mandatory", "none"));
#ifdef HAVE_TLS
	ConfigSection* sub = section->AddSection("tls", "TLS information", MyConfig::OPTIONAL);
	sub->AddItem(new ConfigItem_string("trust_file", "CA certificate file for TLS", " "));
	sub->AddItem(new ConfigItem_string("crl_file", "CA certificate file for TLS", " "));
	sub->AddItem(new ConfigItem_string("cert_file", "Server certificate file for TLS"));
	sub->AddItem(new ConfigItem_string("key_file", "Server key file for TLS"));
	sub->AddItem(new ConfigItem_string("priority", "Priority list for ciphers, exchange methods, macs and compression methods", "NORMAL"));
#endif
}

Minbif::~Minbif()
{
	delete server_poll;
	remove_pidfile();
}

void Minbif::remove_pidfile(void)
{
	if(pidfile.empty())
		return;
	std::ifstream fp(pidfile.c_str());
	if(!fp)
		return;
	int i;
	fp >> i;
	if(i != getpid())
		return;

	unlink(pidfile.c_str());

}

void Minbif::usage(int argc, char** argv)
{
	std::cerr << "Usage: " << argv[0] << " [OPTIONS]... CONFIG_PATH" << std::endl << std::endl;
	std::cerr << "Options:" << std::endl;
	std::cerr << "  -h, --help             Display this notice" << std::endl;
	std::cerr << "  -v, --version          Version of minbif" << std::endl;
	std::cerr << "  -m, --mode=MODE        Mode to run (MODE is a number)" << std::endl;
	std::cerr << "  -p, --pidfile=PIDFILE  Path to pid file" << std::endl;
}

void Minbif::version(void)
{
	for(size_t i = 0; infotxt[i] != NULL; ++i)
		std::cout << infotxt[i] << std::endl;
}

int Minbif::main(int argc, char** argv)
{
	static struct option long_options[] =
	{
		{ "pidfile",       1, NULL, 'p' },
		{ "help",          0, NULL, 'h' },
		{ "version",       0, NULL, 'v' },
		{ "mode",          1, NULL, 'm' },
		{ NULL,            0, NULL, 0   }
	};
	int option_index = 0, c;
	int mode = -1;
	while((c = getopt_long(argc, argv, "m:p:hv", long_options, &option_index)) != -1)
		switch(c)
		{
		case 'm':
			mode = atoi(optarg);
			break;
		case 'h':
			usage(argc, argv);
			return EXIT_SUCCESS;
			break;
		case 'v':
			version();
			return EXIT_SUCCESS;
			break;
		case 'p':
		{
			std::ifstream fi(optarg);
			if(fi)
			{
				std::cerr << "It seems that minbif is already launched. Perhaps try to erase file " << optarg << std::endl;
				fi.close();
				return EXIT_FAILURE;
			}
			pidfile = optarg;
			break;
		}
		default:
			return EXIT_FAILURE;
			break;
		}

	if(optind >= argc)
	{
		std::cerr << argv[0] << ": wrong argument count" << std::endl;
		usage(argc, argv);
		return EXIT_FAILURE;
	}

	try
	{
		struct rlimit rlim;
		if(!getrlimit(RLIMIT_CORE, &rlim) && rlim.rlim_cur != RLIM_INFINITY)
		{
			rlim.rlim_cur = RLIM_INFINITY;
			rlim.rlim_max = RLIM_INFINITY;
			setrlimit(RLIMIT_CORE, &rlim);
		}

		if(!conf.Load(argv[optind]))
		{
			b_log[W_ERR] << "Unable to load configuration, exiting..";
			return EXIT_FAILURE;
		}
		b_log.setLoggedFlags(conf.GetSection("logging")->GetItem("level")->String(), conf.GetSection("logging")->GetItem("to_syslog")->Boolean());

		/* Set users directory path and if I have rights to write in. */
		im::IM::setPath(conf.GetSection("path")->GetItem("users")->String());

		if (mode < 0)
			mode = conf.GetSection("irc")->GetItem("type")->Integer();

		server_poll = ServerPoll::build((ServerPoll::poll_type_t)mode, this);
		b_log.setServerPoll(server_poll);

		if(!pidfile.empty())
		{
			std::ofstream fo(pidfile.c_str());
			if(!fo)
			{
				std::cerr << "Unable to create file '" << pidfile << "': " << strerror(errno) << std::endl;
				return EXIT_FAILURE;
			}
			fo << getpid() << std::endl;
			fo.close();
		}
		sighandler.setApplication(this);

		g_thread_init(NULL);
		loop = g_main_new(FALSE);
		g_main_run(loop);

		return EXIT_SUCCESS;
	}
	catch(MyConfig::error_exc &e)
	{
		b_log[W_ERR] << "Error while loading:";
		b_log[W_ERR] << e.Reason();
	}
	catch(im::IMError &e)
	{
		b_log[W_ERR] << "Unable to load IM settings: " << e.Reason();
	}
	catch(ServerPollError &e)
	{
		b_log[W_ERR] << "Unable to load the server poll.";
	}

	return EXIT_FAILURE;
}

void Minbif::rehash()
{
	if(!conf.Load())
	{
		b_log[W_ERR] << "Unable to load configuration, exiting..";
		quit();
	}
	b_log.setLoggedFlags(conf.GetSection("logging")->GetItem("level")->String(), conf.GetSection("logging")->GetItem("to_syslog")->Boolean());
	if(server_poll)
		server_poll->rehash();
}

void Minbif::quit()
{
	g_main_quit(loop);
}

int main(int argc, char** argv)
{
	Minbif minbif;
	return minbif.main(argc, argv);
}
